Clipboard Ozone implementation

msisov opened the pull request:
This is a first set of patches, which implement clipboard
support for the Ozone platforms.

The flow is asynchronous and based on Requests. That is,
whenever a ClipboardOzone receives a Read/Write/GetMime
call, it forwards it to the AsyncClipboardOzone, which
then creates a request (used for internal usage and holding
data filled by the ClipboardDelegate), calls to the delegate
and start an abort timer to make sure the request is not stalled.

What is more the clipboard data is cached and removed only when another
chunk of data iss written to a system clipboard. And whenever the chromium browser is the
selection owner, the cached data is used.

msisov committed:
Clipboard: move various predefined FormatTypes to common file.

Many clipboard implementations use the predefined FormatTypes
from the clipboard_aurax11. Thus, to make clipboard ozone use
them as well, move those into a separate clipboard_linux
file and include it when necessary.

Change-Id: Iec6b6aacecc567a35534a170fb50afff7e978140

msisov committed:
[ozone] Implement Clipboard for Ozone platforms.

This is a first patch, which implements basic text
support for the Ozone platforms.

The flow is asynchronous and based on Requests. That is,
whenever a ClipboardOzone receives a Read/Write/GetMime
call, it forwards it to the AsyncClipboardOzone, which
then creates a request (used for internal usage and holding
data filled by the ClipboardDelegate), calls to the delegate
and start an abort timer to make sure the request is not stalled.

What is more the clipboard data is cached and removed only when another
chunk of data iss written to a system clipboard. And whenever the chromium browser is the
selection owner, the cached data is used.

Change-Id: I0101aebe47cf2cac666f0434b6be2a9a11e2418c

msisov committed:
[ozone/wayland] Clipboard: integrate Ozone/Wayland and ClipboardOzone

Change-Id: I3b1253f38681e8555c762eb4560df9fbe34e106f

tonikitoo reviewed (commented):
nitpicking a bit, maybe clipboard_aura.cc or clipboard_auralinux.cc (rather than _linux)?
but lets try to upstream this and hopefully hear from them.

Should this commit include any associated include adaptations?

tonikitoo commented on ui/base/BUILD.gn:
```diff
@@ -650,12 +650,21 @@ jumbo_component("base") {
-      ]
+    if (is_desktop_linux) {
+      if (use_x11 || use_ozone) {
+        sources += [ "clipboard/clipboard_linux.cc" ]
```
will this file be needed for say ozone/win (something google was upstreaming at some point)?

msisov replied:
well, this is not platform specific and can be used with mac or win in theory.

tonikitoo reviewed (approved):
good thing is that it does not change the ozone's clipboard impl, so I can apply my local patches for middle-click paste as a follow up.

r+

jkim-julie commented on ui/base/clipboard/clipboard.h:
```diff
@@ -31,6 +31,10 @@
 #endif
 
+#if defined(USE_OZONE)
+#include "ui/ozone/public/clipboard_delegate.h"
```
nit: It would be great if we could use the forward declaration here?

msisov replied:
it will need another set of if defs, which I don't want to add.

jkim-julie commented:
lgtm!

msisov committed:
[ozone] Add clipboard testing support.

Mock out clipboard delegate and provide a simple mocked
system clipboard implementation to test clipboard_ozone

Change-Id: I4d65e1496930cb14a32396be974e8f84353288d0
